src/gsym/parser.rs

```diff
@@ -178,25 +178,26 @@ impl<'a> GsymContext<'a> {
     /// # Safety
     ///
     /// The code will crash with an invalid offset.
-    pub fn get_str(&self, off: usize) -> &str {
-        assert!(off < self.str_tab.len());
+    pub fn get_str(&self, off: usize) -> Option<&str> {
+        if off >= self.str_tab.len() {
+            return None;
+        }
 
         // Ensure there is a null byte.
         let mut null_off = self.str_tab.len() - 1;
         while null_off > off && self.str_tab[null_off] != 0 {
             null_off -= 1;
         }
         if null_off == off {
-            return "";
+            return Some("");
```

Comment:
should this be None? If we treat \0 as optional, then other alternative seems to be returning a string from off to the end of str_tab?

Reply:
I wouldn't think so: we found a NUL byte at the provided offset. That means that there is a valid string, but it is empty. Whether an empty string has meaning to the caller is a different question, but I am not convinced that returning `None` instead is the right thing.

> If we treat \0 as optional

I am not sure why we would do that? I haven't checked the standard but unless it explicitly states that doing so is valid, I don't think we should go down that road.

In any event, this change is meant to be semantic-preserving, so I think returning `Some("")` is the right call. Let me know if you disagree.

Reply:
`null_off == off` condition only means that above `while null_off > off` didn't find any NUL byte *after* off, but we never check that str_tab[off] is 0

Changing above to roughly:
```
while null_off >= off && ... {
   null_off -= 1;
}
if null_off < off {
    return None;
}
```

makes sense taking into account what you are saying above.

My main concern is that for some error conditions we return None, while for others we return Some(""). It seems fragile and inconsistent. Might be completely harmless, but also can bite us in the future.

Reply:
Sounds good. I will unify the string decoding functionality eventually, in which case we should probably use the [`extract_string`](https://github.com/libbpf/blazesym/blob/master/src/util.rs#L100) equivalent and this problem will go away. Okay?

Reply:
sure